### PR TITLE
Added eagain_not_error and noblock config options

### LINKED ISSUE #12: Allow ZMQ::DONTWAIT setting
ZMQ::DONTWAIT: Logstash output plugin does not wait until ZMQ has sent the event. The send is non blocking.

### PR DESCRIPTION
**Pull request by breml**
When using ZMQ::SNDTIMEO with ZeroMQ logstash output
this will produce error messages with return code ZMQ::EAGAIN.
The eagain_not_error configuration parameter allows to ignore these
error messages.
Without ZMQ::SNDTIMEO ZeroMQ logstash output will not respond to
Ctrl+C interrupt if in send_string and there is no connection
to the peer (PUSHPULL).

If noblock is set to true, send_string will use ZMQ::DONTWAIT.
This allows to send without blocking in PUSHPULL.

Fixes #12

### CODE REVIEW

**Comment by ghost** [comment]
Jenkins standing by to test this. If you aren't a maintainer, you can ignore this comment. Someone with commit access, please review this and clear it for Jenkins to run; then  say 'jenkins, test it'.

**Comment by breml** [comment]
@purbon You are kidding me, right?

There are no tests at all at the moment:
https://github.com/logstash-plugins/logstash-output-zeromq/blob/master/spec/outputs/zeromq_spec.rb

File: lib/logstash/util/zeromq.rb [hunk 1]
```diff
@@ -21,8 +21,8 @@ def setup(socket, address)
     @logger.info("0mq: #{server? ? 'connected' : 'bound'}", :address => address)
   end
 
-  def error_check(rc, doing)
-    unless ZMQ::Util.resultcode_ok?(rc)
+  def error_check(rc, doing, eagain_not_error = false)
```
Comment (breml) [inline lib/logstash/util/zeromq.rb]: Because function signature of error_check is changed and util/zeromq.rb is present in zeromq input and output plugin, this PR has a dependecy to logstash-plugins/logstash-input-zeromq/pull/18.

**Comment by breml** [comment]
This plugin does not have any tests at the moment and I am not able to provide all the tests for this plugin.

**Comment by purbon** [comment]
@breml 

> You are kidding me, right?

I'm not kidding :smile: 

is good to have test, and never is late to have them. We're trying hard to improve the different plugins quality and test coverage. 

I will try to provide some scaffolding for tests here you might pick up. before I can do that, another way would be other people can test this in real life, so please dear community! let us know if this change works :+1:

**Comment by breml** [comment]
I did not intend to offend you.

I would be very happy if you could provide some scaffolding for the tests.

**Comment by purbon** [comment]
> I did not intend to offend you.

no offence taken man! much appreciate all your efforts here!! you are staying everywhere :+1:, I'm sorry however it takes some time to merge all things, so many thing to do in some small hours.

**Comment by breml** [comment]
@purbon Do you have some scaffolding for the tests for me?

**Comment by suyograo** [comment]
@avishai-ish-shalom would be awesome if you could review this patch. We can then merge https://github.com/logstash-plugins/logstash-input-zeromq/pull/18 as @breml mentioned

**Comment by nukemberg** [comment]
I'm putting this PR on hold until logstash-mixin-zeromq is integrated

**Comment by nukemberg** [comment]
Ok, i'm finally read to integrate this. @breml, what is the merit of non-blocking mode? if it's crucial to have ctrl+c working, should we even allow blocking mode? is there an alternate way to have ctrl+c working?

**Comment by breml** [comment]
@avishai-ish-shalom You work quite hard on the ZeroMQ Logstash plugins today. Thanks a lot.

If I remember correctly, the current behavior is to block on sending. Because I did not want to change the default behavior, I implemented non-blocking as a new flag.
If ZeroMQ is unable to setup the connection, the send operation blocks in such way that Logstash does no longer respond correctly to ctrl+c (maybe to other signals as well). This is ugly in my opinion.
But, if we use only the non-blocking mode, this plugin has to handle  EAGAIN error properly, otherwise events may be lost. 

In general, Logstash is designed such that output-plugins, unable to process their events, will stop the whole queue within Logstash from being processed. This is a design decision, made very early in the development of Logstash. Consequently, the normal mode of operation for the ZeroMQ output plugin should be blocking (but still responding to ctrl+c).

On the other hand, there are situation, where the user is willing to accept some loss of events to prevent the whole queue being blocked.

Our use case for non blocking mode is the following:

We use ZeroMQ for transport of events from Logstash to Logstash over internet-VPN. In our case we keep all events (full details) on the sender side (e.g. elasticsearch), but for statistical reasons we do want to have some aggregated values on the receiving end as well. In our use case it is better to have some data loss than to block the whole Logstash queue.
We use the High Water Mark (HWM) feature of ZeroMQ to buffer events and let ZeroMQ compensate for small to medium sized interrupts on the VPN. Because we use HWM with non blocking sending, Logstash does not get blocked, even if our VPN has an interrupt. 

Bottom line: I think the ZeroMQ output plugin should allow both modes, blocking and non blocking, because they serve different requirements.
We should further improve the blocking mode to be still responsive to ctrl+c.

I hope these explanations are clear. I am on holidays this week. I will followup on this next week.

**Comment by nukemberg** [comment]
The ZeroMQ topology suitable for lossy sending is `pubsub` and it should work well in blocking mode. `pushpull` socket isn't designed for lossy messaging, even in non-blocking mode.

As far as termination behaviour is concerned, I think using timeout and/or closing the zeromq context may achieve this.

**Comment by breml** [comment]
@avishai-ish-shalom The pubsub topology is indeed suitable for lossy sending, BUT it really only best effort, which means even small network interrupts will lead to loss of log messages. This is not good enough for us (our use case), because we do observe quite frequent small interrupts on our internet based VPN connection.

For us the solution with pushpull topology combined with the high water mark allows us to easily compensate for small to medium interrupts (up to half an hour) and we would be quite unhappy to loose this option.

In my opinion the user of the ZeroMQ Logstash plugins should have both options: blocking and non-blocking.

**Comment by breml** [comment]
@avishai-ish-shalom Could you please tell me, how we can precede with this PR?

CC: @suyograo

**Comment by nukemberg** [comment]
@breml this relates to the shutdown issues on logstash-mixin-zeromq. Also, wouldn't it be better to use zeromq poller for the non-blocking?

**Comment by breml** [comment]
@avishai-ish-shalom What do you exactly mean by using the zeromq poller? How would this fit into the current zeromq output plugin?

**Comment by nukemberg** [comment]
I'm thinking perhaps we should just ditch the blocking mode and have only a poller based non-blocking mode. That would require a rewrite of course.
